[scroll-animations] Implement animation-range calculation

nmoucht opened the pull request:
#### 660d658f3b685c7d41c8ba8a1e93c386f17a80b3
<pre>
[scroll-animations] Implement animation-range calculation
<a href="https://bugs.webkit.org/show_bug.cgi?id=280848">https://bugs.webkit.org/show_bug.cgi?id=280848</a>
<a href="https://rdar.apple.com/137232979">rdar://137232979</a>

Reviewed by Antoine Quint.

Implement basic animation-range calculation. This needs to be hooked up via
WebAnimation::currentTime. Will also need some work to parse TimelineRangeValue
to TimelineRange.

* Source/WebCore/animation/AnimationTimeline.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeScrollTimelineData const):
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::computeScrollTimelineData):
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::computeViewTimelineData const):
(WebCore::ViewTimeline::startOffset const):
(WebCore::ViewTimeline::endOffset const):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: <a href="https://commits.webkit.org/284732@main">https://commits.webkit.org/284732@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d459098213fa5a2796464b63b158b7712ca58df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70328 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49734 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74417 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21506 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72445 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57534 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21346 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55732 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14204 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73394 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45251 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60629 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36194 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41910 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18064 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19867 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63841 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18408 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76136 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14554 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17642 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63446 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14596 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60695 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63384 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11423 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5056 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10764 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45536 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/303 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46610 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47887 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46352 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/AnimationTimeline.h:
```diff
@@ -36,6 +37,17 @@ namespace WebCore {
+
+struct TimelineRange {
+    SingleTimelineRange rangeStart;
+    SingleTimelineRange rangeEnd;
```
I think we can omit `range` from the property names.
```suggestion
    SingleTimelineRange start;
    SingleTimelineRange end;
```

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimeline.h:
```diff
@@ -36,6 +37,17 @@ namespace WebCore {
 
+enum class NamedTimelineRange { Normal, Omitted, Cover, Contain, Entry, Exit, EntryCrossing, ExitCrossing };
+struct SingleTimelineRange {
+    NamedTimelineRange range { NamedTimelineRange::Normal };
```
We could use `name` here and avoid the confusing calls to `range.range`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimeline.h:
```diff
@@ -36,6 +37,17 @@ namespace WebCore {
 class AnimationTimelinesController;
 class WebAnimation;
 
+enum class NamedTimelineRange { Normal, Omitted, Cover, Contain, Entry, Exit, EntryCrossing, ExitCrossing };
```
We could call this `TimelineRangeName` to match the spec type `<timeline-range-name>`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -166,11 +166,22 @@ ScrollTimeline::Data ScrollTimeline::computeScrollTimelineData() const
+    auto computeRangeStart = [&](SingleTimelineRange range) -> float {
+        ASSERT(range.range == NamedTimelineRange::Normal || range.range == NamedTimelineRange::Omitted);
+        return floatValueForLength(range.offset, maxScrollOffset);
+    };
```
Not sure we gain anything using lambdas here since these functions are only used one so using a parameter seems pointless.

nmoucht replied:
Right yeah lambdas aren't necessary for this one.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -254,22 +253,57 @@ ViewTimeline::Data ViewTimeline::computeViewTimelineData() const
 
-    auto coverRangeStart = subjectOffset - scrollContainerSize + insetEnd.value();
-    auto coverRangeEnd = subjectOffset + subjectSize - insetStart.value();
+    auto computeRangeStart = [&](SingleTimelineRange range) -> float {
```
Is the explicit return type required here?

nmoucht replied:
It lets us avoid needing to cast the zero to a float.

graouts replied:
Couldn't we just use `0.f`?

nmoucht replied:
Sure.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -254,22 +253,57 @@ ViewTimeline::Data ViewTimeline::computeViewTimelineData() const
+            return 0;
+        }
+    };
+    auto computeRangeEnd = [&](SingleTimelineRange range) -> float {
```
Ditto about the return type.

nmoucht replied:
Same as above.

nmoucht replied:
Fixed.

graouts reviewed (changes requested):
I'm not sure `NamedTimelineRange` and `SingleTimelineRange` belong in `AnimationTimeline.h`. These might be better placed in their own header file or within `ScrollTimeline.h`.

graouts commented on Source/WebCore/animation/AnimationTimeline.h:
```diff
@@ -26,6 +26,7 @@
 
 #pragma once
 
+#include "Length.h"
```
Is this necessary?

nmoucht replied:
Removed.

graouts commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -154,8 +154,11 @@ ScrollableArea* ScrollTimeline::scrollableAreaForSourceRenderer(RenderElement* r
 }
 
-ScrollTimeline::Data ScrollTimeline::computeScrollTimelineData() const
+ScrollTimeline::Data ScrollTimeline::computeScrollTimelineData(TimelineRange range) const
```
Could this be a `const` reference?

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRange.h:
```diff
@@ -0,0 +1,43 @@
+struct SingleTimelineRange {
+    TimelineRangeName name { TimelineRangeName::Normal };
+    Length offset;
+};
```
I would put the `enum` within the `struct` and just call it `Name`, assuming that's possible, but I think you prefer not to do that.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -217,7 +217,7 @@ RenderBox* ViewTimeline::sourceRenderer() const
 }
 
-ViewTimeline::Data ViewTimeline::computeViewTimelineData() const
+ViewTimeline::Data ViewTimeline::computeViewTimelineData(TimelineRange range) const
```
Can we use a `const` reference here?

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-animations] Implement animation-range calculation
https://bugs.webkit.org/show_bug.cgi?id=280848
rdar://137232979

Reviewed by Antoine Quint.

Implement basic animation-range calculation. This needs to be hooked up via
WebAnimation::currentTime. Will also need some work to parse TimelineRangeValue
to TimelineRange.

* Source/WebCore/animation/AnimationTimeline.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeScrollTimelineData const):
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::computeScrollTimelineData):
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::computeViewTimelineData const):
(WebCore::ViewTimeline::startOffset const):
(WebCore::ViewTimeline::endOffset const):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: https://commits.webkit.org/284732@main
